NextUp-comedy/api-nc
# Update emailing service to sendgrid

## In This PR:

- [x] switch from [nodemailer](https://nodemailer.com/about/) to [sendgird](https://sendgrid.com/) service.
- [x] update email templates.


* **effected endpoints**:
    1.  `POST`: `api/v1/auth/signup/` - as User
    2. `POST`: `api/v1/auth/login/` - both User & Admin 
    3. `GET`: `api/v1/user/waiting-list/` - as Admin
    4. `POST`: `api/v1/user/approve/:userId` - as Admin
    5. `POST`: `api/v1/user/reject/:userId` - as Admin
    6. `POST`: `api/v1/auth/forget-password` - as User
    7. `POST`:`api/v1/auth/reset-password` - as a User. 
---
### Pricing:
- with our free plan now, we can send up to 100 emails/day.
- exceeding that free amount, to send between [1000 - 50,000] email/month, will cost us/them ~$19.95/monthly.
for more [see pricing section](https://sendgrid.com/pricing/).
-----
### Email Templates:
- [x] the user should be expecting 4 different email types from us:
  - Verification email
  - Approval Email
  - Rejection Email
  - Reset password email
<div align="center">
<img src ="https://user-images.githubusercontent.com/56412800/180400838-7f051fc2-0195-4013-af97-2192e36c6fb6.png" height =230 width=150 />
<img src ="https://user-images.githubusercontent.com/56412800/180402584-0fc75d49-81f4-45ba-8b18-4747fe77cc07.png" height =230 width=150 />
<img src ="https://user-images.githubusercontent.com/56412800/180403075-2ea20423-cc51-4999-9d3a-c58620daf6f9.png" height =230 width=150 />
<img src ="https://user-images.githubusercontent.com/56412800/180403835-47ff2222-8e44-47ae-92d9-a7bbba60a6e6.png" height =230 width=150 />
</div>

----
## Testing
#### Env Config
1. setup your backend environment.
   -  <details>
      <summary>create `.env.development` with these variables.</summary>
      
      ```
        PORT=5000
        DATABASE_URL = postgres://<USERNAME>:<PASSWORD>@localhost:5432/nextup_comedy_dev
        TEST_DB_URL = postgres://<USERNAME>:<PASSWORD>@localhost:5432/nextup_comedy_test
        JWT_SECRET=<IT_IS_CALLED_SECRET_FOR_A_REASON>
        NODEMAILER_APP_MAIL = <IT_IS_CALLED_SECRET_FOR_A_REASON>
        NODEMAILER_MAIL_PASSWORD = <IT_IS_CALLED_SECRET_FOR_A_REASON>
        CLIENT_BASE_URL = <IT_IS_CALLED_SECRET_FOR_A_REASON>
        SERVER_BASE_URL = <IT_IS_CALLED_SECRET_FOR_A_REASON>
        
        SENDGRID_ADMIN_EMAIL = <IT_IS_CALLED_SECRET_FOR_A_REASON>
        SENDGRID_API_KEY = <IT_IS_CALLED_SECRET_FOR_A_REASON>
        SENDGRID_VERIFICATION_TEMPLATE_ID = <IT_IS_CALLED_SECRET_FOR_A_REASON>
        SENDGRID_RESET_PASSWORD_TEMPLATE_ID = <IT_IS_CALLED_SECRET_FOR_A_REASON>
        SENDGRID_APPROVAL_TEMPLATE_ID = <IT_IS_CALLED_SECRET_FOR_A_REASON>
        SENDGRID_REJECTION_TEMPLATE_ID = <IT_IS_CALLED_SECRET_FOR_A_REASON>
        NEXTUP_COMEDY_SUPPORT_EMAIL = <IT_IS_CALLED_SECRET_FOR_A_REASON>
        
        CLIENT_ID=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXx
        CLIENT_SECRET=XXXXXXXXXXXXXXXXx
      ```
      </details>
  - `$ npm run build:db` >`$ npm run insert:db`
  - `$ npm install `:
     -  `$ npm run dev`
     - OR : create `.env` > `$ npm run build` > `$ npm run start`
 2. For a better dev experience, run your front repo too:
    - clone `https://github.com/NextUp-comedy/front-nc`
    - setup your `.env` file with:
      - <details>
         <summary>front env vars</summary>
       
         ```
            REACT_APP_CLIENT_ID = <XXXXXXXXXXX>
            REACT_APP_CLIENT_SECRET=XXXXXXXXXXXXXXXX
            SKIP_PREFLIGHT_CHECK=XXXX
         ```
        </details
    - `$ npm install` > `$ npm start`
----
#### Test this PR:
 1. Go to `api/v1/auth/signup/`:
      - Request "create user DTO"
       ```ts
       {
        "username": "your new name",
        "email": "valid_email@gmail.com",
        "password": "not_less_than_6_char"
       }
    ```
    - expected Response:
    ```ts
      {  "message": ""Check your email"   } // check your Email. " Spam / Primary / Promotions/ Updates / wait for half a min max if it wasn't there yet & make sure you've written your email correctly. 
     ```
2. See how the whole journey goes in this video 👇🏽 : 
   https://www.loom.com/share/bd8fbc4387b84a398cbf42d447d1b581
